### DISCUSSION TITLE
ESP32-WROOM-32

### DISCUSSION THREAD

**Discussion by mattaipm** [arendst/Tasmota#22201] [Support request]
I want to compile my own batch with A4988
to ESP32-WROOM-32(ESP32WROOM32)
But I can't set the configuration correctly in platformio_override.ini
I get

After upload in https://tasmota.github.io/install/ 

ets Jul 29 2019 12:21:46

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
ets Jul 29 2019 12:21:46

**Comment by sfromis** [comment]
Looks like you may have selected the wrong build environment. As a starting point, you should build `tasmota32` instead of `tasmota` (which is for ESP8266).

And when flashing ESP32 (instead of OTA upgrading), make sure to use the `.factory` variant of the binary.

**Comment by mattaipm** [comment]
I tried on tasmota32 - the effect is very similar.
I made the changes in the previously mentioned *.ini file
Load the generated file via the page or esptool (what parameters, offset)?

**Reply by sfromis** [reply]
With the Web Installer, did you make sure to follow **both** points of what I wrote? The last one is very important.

Anyway, the offset is used by the Web Installer is `0x010000`

**Comment by mattaipm** [comment]
similarly with older ESPs.
First you had to send .factory.bin - it zeroed and initially configured the flash content
Then send the target FW with an offset, or is it similarly two-step?

-----

I uploaded the factory via web/install
After restarting I flashed my firmware via the device panel
SAFEBOOT
ESP32-DevKit
Tasmota
Upload Failed

Upload failed. Enable logging 3

What am I doing wrong?

**Reply by sfromis** [reply]
Hard to tell as you are not detailed about what you do, like if you are following the suggestion about looking in the log. And "my firmware" is no info about what *exactly* that is.

Also unclear about what you ask about with "two-step". Flashing a `.factory` binary is very similar to flashing ESP8266, a single-step operation, where you normally would include erase in this.

The `.factory` variant is the **full** Tasmota, also when you pick this from your build. Do look in `build_output` for files, do not pick those intermediary files with names starting with `firmware`.

**Comment by mattaipm** [comment]
I programmed a lot of esp 12e/f
I recompiled with the A4988 option and that's it. Everything works fine.
I changed the layout and wanted to put it into this version.
But there is a problem



After compilation I get the file <xxxx>.bin
I always uploaded it and that's it.

In this case I have in the logs on www /install what I gave at the beginning

**Reply by sfromis** [reply]
The booting errors tells nothing more that you did something wrong, and therefore - if you want suggestions, you need to post specific details of the exact steps, including full file names, not vague generalities like including some build time option. As there are more ways of flashing ESP32 compared to ESP8266, it gets more important to do the steps just right, which the web installer helps you with.

What I meant by logs was if you followed the specific suggestion with the error message:
> Upload failed. Enable logging 3

Before trying again with the web installer. Of course, reading the log messages yourself is also important.

Note that when you first flash a precompiled binary via the web installer, and get it working to the point of trying to use the Upgrade function in the web UI, then you need point it at a file name like `tasmota32.bin`, not the `tasmota32.factory.bin` needed for initial flashing via USB.

**Comment by mattaipm** [comment]
[20240927-1250.firmware.bin.zip](https://github.com/user-attachments/files/17163036/20240927-1250.firmware.bin.zip)

All ready bin files via www /install are uploaded correctly
I can't upload my compiled in docker-tasmota environment using the standard procedure.

Everything works on tasmota, switching to tasmota32 in the .ini file the resulting file doesn't load correctly
I don't see any hint because apart from the message I provided I don't see anything else on the www /installer page.

Where can I get information about problems with compilation - maybe there is a problem at this stage.

**Reply by sfromis** [reply]
The file you posted is absolutely built the wrong way, as it is for ESP8266 instead of ESP32. You have not made the switch to ESP32 builds the right way.

Also, I just told you to **not use** the files with names starting with `firmware`, but find the right files in the `build_output` folder. That **should** be names like `tasmota32.bin` and `tasmota32.factory.bin`. Certainly not ESP8266 files like `tasmota.bin` or `tasmota.bin.gz`.

**Comment by mattaipm** [comment]
![Screenshot_20240927_145223](https://github.com/user-attachments/assets/d62d9058-794d-4249-a232-7a5dc705cfbb)

![Screenshot_20240927_144954](https://github.com/user-attachments/assets/bae6f6e8-1736-4b1d-8266-e68058274cc9)
Thank you very much 

The attached file is a copy of the file
docker-tasmota/Tasmota/build_output/firmware/tasmota-PL.bin

**Reply by sfromis** [reply]
When you're building `tasmota-PL`, you are not using the *default* you can (or not) set in the file, but instead selecting a build environment when you start the build. That's where you should select `tasmota32` **instead of** now selecting `tasmota-PL`. I'd even say that this way is actually preferable instead of modifying the file to select one. Just make sure that you also make the right selection when you start the build.

You likely have something like `-e tasmota-PL` on the command line when you build, but you are still not showing your basic steps.

**Comment by mattaipm** [comment]
Sorry, you're absolutely right :-)
I forgot that there are parameters there, I have a fairly automated environment. I'm an electronics engineer, not a programmer, and I make it as easy as I can.

docker run -ti --rm -v $(pwd)/Tasmota:/tasmota -u $UID:$GID docker-tasmota -e tasmota-PL

**Comment by mattaipm** [comment]
docker run -ti --rm -v $(pwd)/Tasmota:/tasmota -u $UID:$GID docker-tasmota -e tasmota32

after compilation

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
wrong chip id 0x63a8
ets Jul 29 2019 12:21:46


[log.txt](https://github.com/user-attachments/files/17179934/log.txt)

**Reply by sfromis** [reply]
Again, you are hiding your basic steps around what **exactly** you flashed and how. Maskes it impossible to spot what you did not do right.

**Comment by mattaipm** [comment]
What should I do to see what is necessary for diagnostics?

**Reply by sfromis** [reply]
Your initial post was about having flashed a completely wrong binary. Now you have the exact same symptoms, and I just explained to you that the important thing is 
> your basic steps around what **exactly** you flashed and how

**Comment by mattaipm** [comment]
My sources **only** get from 
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
[remote "origin"]
        url = https://github.com/tasmota/docker-tasmota
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master

**Comment by jo-mueller** [comment]
@sfromis thanks a bunch for your explanations above, I think I ran into precisely the issue you lay out to resolve. I did not know that I had to flash my esp32 with the `tasmota32.factory.bin` in order to continue.

Just to double-check: Is this command building the correct binary for an ESP32 WROOM 32?

```bash
pio run -e tasmota32
```

or, respectively

```bash
pio run -e tasmota32c3
```

Just to make sure I get the process right:

- I compile the binaries (I use gitpod) with platformio
- I don't the built binaries from the `build_output` folder
- I flash the esp32 using any flash tool out there with the `tasmota32.factory.bin` (this works)
- I upgrade the esp32 with the `tasmota32.bin` over the air?

**Reply by sfromis** [reply]
The [Tasmota Web Installer](https://tasmota.github.io/install/) is recommended, also when you did the build.

You only need the OTA upgrade with `tasmota32.bin` if you've built a new version to upgrade to, as it is already included in flashing `tasmota32.factory.bin` (which in principle is only used once first time). You can also OTA upgrade to `tasmota32.bin` from a different Tasmota build for the same ESP32 variant.

**Reply by jo-mueller** [reply]
> which in principle is only used once first time

I think that's the part that eluded me and took me waaaay too long to realize.
